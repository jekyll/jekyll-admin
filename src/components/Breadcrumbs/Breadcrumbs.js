import React from 'react';
import PropTypes from 'prop-types';
import { Breadcrumb } from 'antd';
import { Link } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';
import styled from 'styled-components';
import { ADMIN_PREFIX } from 'config';

const Breadcrumbs = ({ root, splat }) => {
  const fullPath = splat ? [root, splat].join('/') : root;
  const breadcrumbs = [];
  fullPath.split('/').reduce((path, dir, index, arr) => {
    const to = `${path}/${dir}`;
    breadcrumbs.push(
      <Breadcrumb.Item key={dir}>
        <Link to={`${ADMIN_PREFIX}${to}`}>
          {index === 0 ? <FormattedMessage id="sidebar.pages" /> : dir}
        </Link>
      </Breadcrumb.Item>
    );
    return to;
  }, '');
  return <StyledBreadcrumb>{breadcrumbs}</StyledBreadcrumb>;
};

Breadcrumbs.propTypes = {
  root: PropTypes.string,
  splat: PropTypes.string,
};

const StyledBreadcrumb = styled(Breadcrumb)`
  margin: 16px 0;
`;

export default Breadcrumbs;
