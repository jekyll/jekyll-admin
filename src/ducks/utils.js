// Action Types
export const SEARCH_CONTENT = 'SEARCH_CONTENT';
export const CLEAR_ERRORS = 'CLEAR_ERRORS';
export const VALIDATION_ERROR = 'VALIDATION_ERROR';

// Actions
export const search = input => ({
  type: SEARCH_CONTENT,
  input,
});

export const clearErrors = () => ({
  type: CLEAR_ERRORS,
});

export const validationError = errors => ({
  type: VALIDATION_ERROR,
  errors,
});

// Reducer
export default function utils(
  state = {
    input: '',
    errors: [],
  },
  action
) {
  switch (action.type) {
    case SEARCH_CONTENT:
      return {
        ...state,
        input: action.input,
      };
    case CLEAR_ERRORS:
      return {
        ...state,
        errors: [],
      };
    case VALIDATION_ERROR:
      return {
        ...state,
        errors: action.errors,
      };
    default:
      return {
        ...state,
        input: '',
      };
  }
}
