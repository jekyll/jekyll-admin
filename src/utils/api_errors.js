export class BadInputError extends Error {
  constructor(message) {
    super();
    this.name = "BadInputError";
    this.message = message || 'Bad input';
    this.stack = (new Error()).stack;
  }
}
