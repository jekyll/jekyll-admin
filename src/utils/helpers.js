import yaml from 'js-yaml';
import _ from 'underscore';

/**
 * Converts the object into YAML string.
 * @param {Object} object
 * @return {String} yaml
 */
export const toYAML = obj =>
  !_.isEmpty(obj) ? yaml.safeDump(obj, { indent: 2 }) : '';

/**
 * Converts the YAML string into JS object.
 * @param {String} string
 * @return {Object} obj
 */
export const toJSON = yamlString => (yamlString ? yaml.load(yamlString) : {});

/**
 * Capitalize the given string.
 * @param {String} string
 * @return {String} string
 */
export const capitalize = string => {
  if (!string) return '';
  return string.charAt(0).toUpperCase() + string.substring(1).toLowerCase();
};

/**
 * Convert the given string into title case format.
 * @param {String} string
 * @return {String} string
 */
export const toTitleCase = string => {
  if (!string) return '';
  return string.replace(/\w\S*/g, txt => {
    return txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase();
  });
};

/**
 * Slugify the given string
 * @param {String} string
 * @return {String} string
 */
export const slugify = string => {
  if (!string) return '';
  str = String(str).toString();
    str = str.replace(/^\s+|\s+$/g, ""); // trim
    str = str.toLowerCase();

    // remove accents, swap ñ for n, etc
    const swaps = {
        '0': ['°', '₀', '۰', '０'],
        '1': ['¹', '₁', '۱', '１'],
        '2': ['²', '₂', '۲', '２'],
        '3': ['³', '₃', '۳', '３'],
        '4': ['⁴', '₄', '۴', '٤', '４'],
        '5': ['⁵', '₅', '۵', '٥', '５'],
        '6': ['⁶', '₆', '۶', '٦', '６'],
        '7': ['⁷', '₇', '۷', '７'],
        '8': ['⁸', '₈', '۸', '８'],
        '9': ['⁹', '₉', '۹', '９'],
        'a': ['à', 'á', 'ả', 'ã', 'ạ', 'ă', 'ắ', 'ằ', 'ẳ', 'ẵ', 'ặ', 'â', 'ấ', 'ầ', 'ẩ', 'ẫ', 'ậ', 'ā', 'ą', 'å', 'α', 'ά', 'ἀ', 'ἁ', 'ἂ', 'ἃ', 'ἄ', 'ἅ', 'ἆ', 'ἇ', 'ᾀ', 'ᾁ', 'ᾂ', 'ᾃ', 'ᾄ', 'ᾅ', 'ᾆ', 'ᾇ', 'ὰ', 'ά', 'ᾰ', 'ᾱ', 'ᾲ', 'ᾳ', 'ᾴ', 'ᾶ', 'ᾷ', 'а', 'أ', 'အ', 'ာ', 'ါ', 'ǻ', 'ǎ', 'ª', 'ა', 'अ', 'ا', 'ａ', 'ä'],
        'b': ['б', 'β', 'ب', 'ဗ', 'ბ', 'ｂ'],
        'c': ['ç', 'ć', 'č', 'ĉ', 'ċ', 'ｃ'],
        'd': ['ď', 'ð', 'đ', 'ƌ', 'ȡ', 'ɖ', 'ɗ', 'ᵭ', 'ᶁ', 'ᶑ', 'д', 'δ', 'د', 'ض', 'ဍ', 'ဒ', 'დ', 'ｄ'],
        'e': ['é', 'è', 'ẻ', 'ẽ', 'ẹ', 'ê', 'ế', 'ề', 'ể', 'ễ', 'ệ', 'ë', 'ē', 'ę', 'ě', 'ĕ', 'ė', 'ε', 'έ', 'ἐ', 'ἑ', 'ἒ', 'ἓ', 'ἔ', 'ἕ', 'ὲ', 'έ', 'е', 'ё', 'э', 'є', 'ə', 'ဧ', 'ေ', 'ဲ', 'ე', 'ए', 'إ', 'ئ', 'ｅ'],
        'f': ['ф', 'φ', 'ف', 'ƒ', 'ფ', 'ｆ'],
        'g': ['ĝ', 'ğ', 'ġ', 'ģ', 'г', 'ґ', 'γ', 'ဂ', 'გ', 'گ', 'ｇ'],
        'h': ['ĥ', 'ħ', 'η', 'ή', 'ح', 'ه', 'ဟ', 'ှ', 'ჰ', 'ｈ'],
        'i': ['í', 'ì', 'ỉ', 'ĩ', 'ị', 'î', 'ï', 'ī', 'ĭ', 'į', 'ı', 'ι', 'ί', 'ϊ', 'ΐ', 'ἰ', 'ἱ', 'ἲ', 'ἳ', 'ἴ', 'ἵ', 'ἶ', 'ἷ', 'ὶ', 'ί', 'ῐ', 'ῑ', 'ῒ', 'ΐ', 'ῖ', 'ῗ', 'і', 'ї', 'и', 'ဣ', 'ိ', 'ီ', 'ည်', 'ǐ', 'ი', 'इ', 'ی', 'ｉ'],
        'j': ['ĵ', 'ј', 'Ј', 'ჯ', 'ج', 'ｊ'],
        'k': ['ķ', 'ĸ', 'к', 'κ', 'Ķ', 'ق', 'ك', 'က', 'კ', 'ქ', 'ک', 'ｋ'],
        'l': ['ł', 'ľ', 'ĺ', 'ļ', 'ŀ', 'л', 'λ', 'ل', 'လ', 'ლ', 'ｌ'],
        'm': ['м', 'μ', 'م', 'မ', 'მ', 'ｍ'],
        'n': ['ñ', 'ń', 'ň', 'ņ', 'ŉ', 'ŋ', 'ν', 'н', 'ن', 'န', 'ნ', 'ｎ'],
        'o': ['ó', 'ò', 'ỏ', 'õ', 'ọ', 'ô', 'ố', 'ồ', 'ổ', 'ỗ', 'ộ', 'ơ', 'ớ', 'ờ', 'ở', 'ỡ', 'ợ', 'ø', 'ō', 'ő', 'ŏ', 'ο', 'ὀ', 'ὁ', 'ὂ', 'ὃ', 'ὄ', 'ὅ', 'ὸ', 'ό', 'о', 'و', 'θ', 'ို', 'ǒ', 'ǿ', 'º', 'ო', 'ओ', 'ｏ', 'ö'],
        'p': ['п', 'π', 'ပ', 'პ', 'پ', 'ｐ'],
        'q': ['ყ', 'ｑ'],
        'r': ['ŕ', 'ř', 'ŗ', 'р', 'ρ', 'ر', 'რ', 'ｒ'],
        's': ['ś', 'š', 'ş', 'с', 'σ', 'ș', 'ς', 'س', 'ص', 'စ', 'ſ', 'ს', 'ｓ'],
        't': ['ť', 'ţ', 'т', 'τ', 'ț', 'ت', 'ط', 'ဋ', 'တ', 'ŧ', 'თ', 'ტ', 'ｔ'],
        'u': ['ú', 'ù', 'ủ', 'ũ', 'ụ', 'ư', 'ứ', 'ừ', 'ử', 'ữ', 'ự', 'û', 'ū', 'ů', 'ű', 'ŭ', 'ų', 'µ', 'у', 'ဉ', 'ု', 'ူ', 'ǔ', 'ǖ', 'ǘ', 'ǚ', 'ǜ', 'უ', 'उ', 'ｕ', 'ў', 'ü'],
        'v': ['в', 'ვ', 'ϐ', 'ｖ'],
        'w': ['ŵ', 'ω', 'ώ', 'ဝ', 'ွ', 'ｗ'],
        'x': ['χ', 'ξ', 'ｘ'],
        'y': ['ý', 'ỳ', 'ỷ', 'ỹ', 'ỵ', 'ÿ', 'ŷ', 'й', 'ы', 'υ', 'ϋ', 'ύ', 'ΰ', 'ي', 'ယ', 'ｙ'],
        'z': ['ź', 'ž', 'ż', 'з', 'ζ', 'ز', 'ဇ', 'ზ', 'ｚ'],
        'aa': ['ع', 'आ', 'آ'],
        'ae': ['æ', 'ǽ'],
        'ai': ['ऐ'],
        'ch': ['ч', 'ჩ', 'ჭ', 'چ'],
        'dj': ['ђ', 'đ'],
        'dz': ['џ', 'ძ'],
        'ei': ['ऍ'],
        'gh': ['غ', 'ღ'],
        'ii': ['ई'],
        'ij': ['ĳ'],
        'kh': ['х', 'خ', 'ხ'],
        'lj': ['љ'],
        'nj': ['њ'],
        'oe': ['ö', 'œ', 'ؤ'],
        'oi': ['ऑ'],
        'oii': ['ऒ'],
        'ps': ['ψ'],
        'sh': ['ш', 'შ', 'ش'],
        'shch': ['щ'],
        'ss': ['ß'],
        'sx': ['ŝ'],
        'th': ['þ', 'ϑ', 'ث', 'ذ', 'ظ'],
        'ts': ['ц', 'ც', 'წ'],
        'ue': ['ü'],
        'uu': ['ऊ'],
        'ya': ['я'],
        'yu': ['ю'],
        'zh': ['ж', 'ჟ', 'ژ'],
        '(c)': ['©'],
        'A': ['Á', 'À', 'Ả', 'Ã', 'Ạ', 'Ă', 'Ắ', 'Ằ', 'Ẳ', 'Ẵ', 'Ặ', 'Â', 'Ấ', 'Ầ', 'Ẩ', 'Ẫ', 'Ậ', 'Å', 'Ā', 'Ą', 'Α', 'Ά', 'Ἀ', 'Ἁ', 'Ἂ', 'Ἃ', 'Ἄ', 'Ἅ', 'Ἆ', 'Ἇ', 'ᾈ', 'ᾉ', 'ᾊ', 'ᾋ', 'ᾌ', 'ᾍ', 'ᾎ', 'ᾏ', 'Ᾰ', 'Ᾱ', 'Ὰ', 'Ά', 'ᾼ', 'А', 'Ǻ', 'Ǎ', 'Ａ', 'Ä'],
        'B': ['Б', 'Β', 'ब', 'Ｂ'],
        'C': ['Ç', 'Ć', 'Č', 'Ĉ', 'Ċ', 'Ｃ'],
        'D': ['Ď', 'Ð', 'Đ', 'Ɖ', 'Ɗ', 'Ƌ', 'ᴅ', 'ᴆ', 'Д', 'Δ', 'Ｄ'],
        'E': ['É', 'È', 'Ẻ', 'Ẽ', 'Ẹ', 'Ê', 'Ế', 'Ề', 'Ể', 'Ễ', 'Ệ', 'Ë', 'Ē', 'Ę', 'Ě', 'Ĕ', 'Ė', 'Ε', 'Έ', 'Ἐ', 'Ἑ', 'Ἒ', 'Ἓ', 'Ἔ', 'Ἕ', 'Έ', 'Ὲ', 'Е', 'Ё', 'Э', 'Є', 'Ə', 'Ｅ'],
        'F': ['Ф', 'Φ', 'Ｆ'],
        'G': ['Ğ', 'Ġ', 'Ģ', 'Г', 'Ґ', 'Γ', 'Ｇ'],
        'H': ['Η', 'Ή', 'Ħ', 'Ｈ'],
        'I': ['Í', 'Ì', 'Ỉ', 'Ĩ', 'Ị', 'Î', 'Ï', 'Ī', 'Ĭ', 'Į', 'İ', 'Ι', 'Ί', 'Ϊ', 'Ἰ', 'Ἱ', 'Ἳ', 'Ἴ', 'Ἵ', 'Ἶ', 'Ἷ', 'Ῐ', 'Ῑ', 'Ὶ', 'Ί', 'И', 'І', 'Ї', 'Ǐ', 'ϒ', 'Ｉ'],
        'J': ['Ｊ'],
        'K': ['К', 'Κ', 'Ｋ'],
        'L': ['Ĺ', 'Ł', 'Л', 'Λ', 'Ļ', 'Ľ', 'Ŀ', 'ल', 'Ｌ'],
        'M': ['М', 'Μ', 'Ｍ'],
        'N': ['Ń', 'Ñ', 'Ň', 'Ņ', 'Ŋ', 'Н', 'Ν', 'Ｎ'],
        'O': ['Ó', 'Ò', 'Ỏ', 'Õ', 'Ọ', 'Ô', 'Ố', 'Ồ', 'Ổ', 'Ỗ', 'Ộ', 'Ơ', 'Ớ', 'Ờ', 'Ở', 'Ỡ', 'Ợ', 'Ø', 'Ō', 'Ő', 'Ŏ', 'Ο', 'Ό', 'Ὀ', 'Ὁ', 'Ὂ', 'Ὃ', 'Ὄ', 'Ὅ', 'Ὸ', 'Ό', 'О', 'Θ', 'Ө', 'Ǒ', 'Ǿ', 'Ｏ', 'Ö'],
        'P': ['П', 'Π', 'Ｐ'],
        'Q': ['Ｑ'],
        'R': ['Ř', 'Ŕ', 'Р', 'Ρ', 'Ŗ', 'Ｒ'],
        'S': ['Ş', 'Ŝ', 'Ș', 'Š', 'Ś', 'С', 'Σ', 'Ｓ'],
        'T': ['Ť', 'Ţ', 'Ŧ', 'Ț', 'Т', 'Τ', 'Ｔ'],
        'U': ['Ú', 'Ù', 'Ủ', 'Ũ', 'Ụ', 'Ư', 'Ứ', 'Ừ', 'Ử', 'Ữ', 'Ự', 'Û', 'Ū', 'Ů', 'Ű', 'Ŭ', 'Ų', 'У', 'Ǔ', 'Ǖ', 'Ǘ', 'Ǚ', 'Ǜ', 'Ｕ', 'Ў', 'Ü'],
        'V': ['В', 'Ｖ'],
        'W': ['Ω', 'Ώ', 'Ŵ', 'Ｗ'],
        'X': ['Χ', 'Ξ', 'Ｘ'],
        'Y': ['Ý', 'Ỳ', 'Ỷ', 'Ỹ', 'Ỵ', 'Ÿ', 'Ῠ', 'Ῡ', 'Ὺ', 'Ύ', 'Ы', 'Й', 'Υ', 'Ϋ', 'Ŷ', 'Ｙ'],
        'Z': ['Ź', 'Ž', 'Ż', 'З', 'Ζ', 'Ｚ'],
        'AE': ['Æ', 'Ǽ'],
        'Ch': ['Ч'],
        'Dj': ['Ђ'],
        'Dz': ['Џ'],
        'Gx': ['Ĝ'],
        'Hx': ['Ĥ'],
        'Ij': ['Ĳ'],
        'Jx': ['Ĵ'],
        'Kh': ['Х'],
        'Lj': ['Љ'],
        'Nj': ['Њ'],
        'Oe': ['Œ'],
        'Ps': ['Ψ'],
        'Sh': ['Ш'],
        'Shch': ['Щ'],
        'Ss': ['ẞ'],
        'Th': ['Þ'],
        'Ts': ['Ц'],
        'Ya': ['Я'],
        'Yu': ['Ю'],
        'Zh': ['Ж'],
    };

    Object.keys(swaps).forEach((swap) => {
        swaps[swap].forEach(s => {
            str = str.replace(new RegExp(s, "g"), swap);
        })
    });
    return str
        .replace(/[^a-z0-9 -]/g, "") // remove invalid chars
        .replace(/\s+/g, "-") // collapse whitespace and replace by -
        .replace(/-+/g, "-") // collapse dashes
        .replace(/^-+/, "") // trim - from start of text
        .replace(/-+$/, "");
};

/**
 * Returns filename from the given path
 * @param {String} path
 * @return {String} filename
 */
export const getFilenameFromPath = path => {
  if (!path) return '';
  return path.substring(path.lastIndexOf('/') + 1);
};

/**
 * Returns extension from the given path or filename
 * @param {String} path or filename
 * @return {String} extension or nil
 */
export const getExtensionFromPath = path => {
  if (!path) return '';
  const filename = getFilenameFromPath(path);
  const index = filename.lastIndexOf('.');

  if (index > 0) {
    return filename.substring(index + 1);
  } else {
    return '';
  }
};

/**
 * returns the uploaded static files that are being overwritten
 * @param {Array} uploadedFiles
 * @param {Array} currentFiles
 * @return {Array} filenames
 */
export const existingUploadedFilenames = (uploadedFiles, currentFiles) => {
  if (
    (uploadedFiles && !uploadedFiles.length) ||
    (currentFiles && !currentFiles.length)
  ) {
    return [];
  }
  const currentFilenames = _.map(currentFiles, cf =>
    getFilenameFromPath(cf.path)
  );
  return _.chain(uploadedFiles)
    .filter(file => currentFilenames.includes(file.name))
    .map(file => file.name)
    .value();
};

/**
 * Given an Event object, prevents the default event
 * from bubbling, if possible.
 * @param {Event} event
 */
export const preventDefault = event => {
  if (event && event.preventDefault) {
    event.preventDefault();
  }
};

/**
 * Given an object, trims every keys and values recursively.
 * @param {Object} object
 * @return {Object} trimmedObject
 */
export const trimObject = object => {
  if (!_.isObject(object)) return object;
  return _.keys(object).reduce((acc, key) => {
    if (typeof object[key] === 'string') {
      try {
        acc[key.trim()] = JSON.parse(object[key].trim());
      } catch (e) {
        acc[key.trim()] = object[key].trim();
      }
    } else {
      acc[key.trim()] = trimObject(object[key]);
    }
    return acc;
  }, Array.isArray(object) ? [] : {});
};

/**
 * Given a resource type and splat, returns a formatted title string.
 * Falsy values are discarded.
 * @param {String} type - Resource type.
 * @param {String} splat - Directory splat for current resource.
 * @param {String} [prefix=''] - Optional string text to be placed in front.
 * @return {String} Empty string or formatted title.
 */
export const getDocumentTitle = (type, splat, prefix = '') => {
  if (!type) return '';
  const label = toTitleCase(type.toString());
  return [prefix, splat, label].filter(Boolean).join(' | ');
};

/**
 * @param {String} directory - Directory splat for current resource.
 * @param {String} filename - Basename of current resource.
 * @return {String} Filename or directory splat joined to the filename.
 */
export const computeRelativePath = (directory, filename) => {
  return directory ? `${directory}/${filename}` : `${filename}`;
};

// omit raw_content, path and empty-value keys in metadata state from front_matter
export const sanitizeFrontMatter = metadata => {
  return _.omit(metadata, (value, key, object) => {
    return key === 'raw_content' || key === 'path' || value === '';
  });
};

export const preparePayload = obj => JSON.stringify(trimObject(obj));
export const getFilenameFromTitle = title => `${slugify(title)}.md`;
