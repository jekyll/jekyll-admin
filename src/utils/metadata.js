import _ from 'underscore';
import cloneDeep from 'lodash/cloneDeep';

/**
 * Returns the metadata of the state with the new empty field. If the field does
 * not exist, returns the original metadata. Does not mutate the given state.
 * @param {Object} state
 * @param {String} namePrefix
 * @return {Object} metadata
 */
export const addField = (state, namePrefix) => {
  let tmpState = cloneDeep(state);
  let field = eval(`tmpState.${namePrefix}`);
  if (field === undefined) return tmpState.metadata;
  if (_.isArray(field)) field.push('');
  else field[`New field ${state.new_field_count}`] = '';
  return tmpState.metadata;
};

/**
 * Returns the metadata of the state with the removed key. If the field does not
 * exist, returns the original metadata. Does not mutate the given state.
 * @param {Object} state
 * @param {String} namePrefix
 * @param {String} key
 * @return {Object} metadata
 */
export const removeField = (state, namePrefix, key) => {
  let tmpState = cloneDeep(state);
  let field = eval(`tmpState.${namePrefix}`);
  if (field === undefined) return tmpState.metadata;
  if (_.isArray(field)) {
    if (key >= field.length) {
      return tmpState.metadata;
    }
    field.splice(key, 1);
  }
  else {
    if (!_.has(field, key)) {
      return tmpState.metadata;
    }
    delete field[key];
  }
  return tmpState.metadata;
};

/**
 * Returns the metadata of the state with the updated key. If the field does not
 * exist or the key already exists, returns the original metadata. Does not
 * mutate the given state.
 * @param {Object} state
 * @param {String} namePrefix
 * @param {String} fieldKey
 * @param {String} newKey
 * @return {Object} metadata
 */
export const updateFieldKey = (state, namePrefix, fieldKey, newKey) => {
  let tmpState = cloneDeep(state);
  let field = eval(`tmpState.${namePrefix}`);
  if (field === undefined) return tmpState.metadata;
  if (_.has(field, newKey)) return tmpState.metadata;
  field = Object.keys(field)
    .reduce((result, current) => {
      if (current == fieldKey) result[newKey] = field[current];
      else result[current] = field[current];
      return result;
    }, {});
  eval(`tmpState.${namePrefix} = field`);
  return tmpState.metadata;
};

/**
 * Returns the metadata of the state with the updated value of given path(nameAttr).
 * If the field does not exist, creates a new field. Does not mutate the given state.
 * @param {Object} state
 * @param {String} nameAttr
 * @param {String} value
 * @return {Object} metadata
 */
export const updateFieldValue = (state, nameAttr, value) => {
  let tmpState = cloneDeep(state);
  eval(`tmpState.${nameAttr} = value`);
  return tmpState.metadata;
};

/**
 * Returns the metadata of the state with the converted type of given path(nameAttr).
 * If the field does not exist, returns the original metadata.
 * Does not mutate the given state.
 * @param {Object} state
 * @param {String} nameAttr
 * @param {String} convertType
 * @return {Object} metadata
 */
export const convertField = (state, nameAttr, convertType) => {
  let tmpState = cloneDeep(state);
  let field = eval(`tmpState.${nameAttr}`);
  if (field === undefined) return tmpState.metadata;
  if (convertType == 'array') field = [''];
  else if (convertType == 'object') {
    let key = `New field ${state.new_field_count}`;
    field = { [key]: '' };
  }
  else field = '';
  eval(`tmpState.${nameAttr} = field`);
  return tmpState.metadata;
};

/**
 * Returns the metadata of the state with the sorted array. Moves the array item to
 * target index, shifts the rest of them. If the given path is not an array,
 * returns the original metadata. Does not mutate the given state.
 * @param {Object} state
 * @param {String} namePrefix
 * @param {Number} srcInd
 * @param {Number} targetInd
 * @return {Object} metadata
 */
export const moveArrayItem = (state, namePrefix, srcInd, targetInd) => {
  let tmpState = cloneDeep(state);
  let arr = eval(`tmpState.${namePrefix}`);
  if (!_.isArray(arr)) return tmpState.metadata;
  arr.splice(targetInd, 0, arr.splice(srcInd, 1)[0]);
  return tmpState.metadata;
};

/**
 * Injects the relevant default front matter fields read from the config file into
 * the provided front matter object
 * @param {Object} config
 * @param {String} path
 * @param {String} type
 * @param {Object} front_matter
 * @return {Object} metadata
 */
export const injectDefaultFields = (config, path, type, front_matter={}) => {
  let defaults;
  try {
    defaults = config.content.defaults;
  } catch (e) {
    return {};
  }
  let metafields = {};
  _.each(defaults, field => {
    const scope = field.scope;
    if ((!scope.type || scope.type == type) && (!scope.path || scope.path == path)) {
      _.extend(metafields, field.values);
    }
  });
  return _.extend(metafields, front_matter);
};
