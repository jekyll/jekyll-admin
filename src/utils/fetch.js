import fetch from 'isomorphic-fetch';
import { addNotification } from '../actions/notifications';
import {
  getErrorMessage,
  getFetchErrorMessage,
  getUpdateErrorMessage,
  getDeleteMessage
} from '../constants/lang';
import { BadInputError } from './api_errors';

/**
 * Fetch wrapper for GET request that dispatches actions according to the
 * request status
 * @param {String} url
 * @param {Object} action_success
 * @param {Object} action_failure
 * @return {Function} dispatch
 */
export const get = (url, action_success, action_failure, dispatch) => {
  return fetch(url)
    .then(res => res.json())
    .then(data => dispatch({
      type: action_success.type,
      [action_success.name]: data
    }))
    .catch(error => {
      dispatch({
        type: action_failure.type,
        [action_failure.name]: error
      });
      dispatch(addNotification(
        getErrorMessage(),
        getFetchErrorMessage(action_success.name),
        'error'
      ));
    });
};

/**
 * Fetch wrapper for PUT request that dispatches actions according to the
 * request status
 * @param {String} url
 * @param {Object} body
 * @param {Object} action_success
 * @param {Object} action_failure
 * @return {Function} dispatch
 */
export const put = (url, body, action_success, action_failure, dispatch) => {
  return fetch(url, {
    method: 'PUT',
    body
  })
  .then(res => res.json())
  .then(data => {
    if (data.error_message){
      throw new BadInputError(data.error_message);
    }
    dispatch({
      type: action_success.type,
      [action_success.name]: data
    });
  })
  .catch(error => {
    dispatch({
      type: action_failure.type,
      [action_failure.name]: error
    });
    let error_message = error.name ==='BadInputError' ? error.message : getUpdateErrorMessage(action_success.name);
    dispatch(addNotification(
      getErrorMessage(),
      error_message,
      'error'
    ));
  });
};

/**
 * Fetch wrapper for DELETE request that dispatches actions according to the
 * request status
 * @param {String} url
 * @param {Object} action_success
 * @param {Object} action_failure
 * @return {Function} dispatch
 */
export const del = (url, action_success, action_failure, dispatch) => {
  return fetch(url, {
    method: 'DELETE'
  })
  .then(data => dispatch({
    type: action_success.type,
    id: action_success.id
  }))
  .catch(error => {
    dispatch({
      type: action_failure.type,
      [action_failure.name]: error
    });
    dispatch(addNotification(
      getErrorMessage(),
      getDeleteMessage(action_success.name),
      'error'
    ));
  });
};
