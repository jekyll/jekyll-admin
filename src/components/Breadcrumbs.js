import React, { Component, PropTypes } from 'react';
import { Link } from 'react-router';
import _ from 'underscore';
import { toTitleCase } from '../utils/helpers';
import { ADMIN_PREFIX } from '../constants';

export default class Breadcrumbs extends Component {

  render() {
    const { splat, type } = this.props;
    // generate links from `splat`
    let base;
    if (type == 'pages') {
      base = `${ADMIN_PREFIX}/pages`;
    } else if (type == 'data files') {
      base = `${ADMIN_PREFIX}/datafiles`;
    } else {
      base = `${ADMIN_PREFIX}/collections/${type}`;
    }
    let links;
    if (splat) {
      const paths = splat.split('/');
      links = _.map(paths, (path, i) => {
        const before = (i == 0) ? '' : paths.slice(0, i).join('/') + '/';
        return {
          href: `${base}/${before}${path}`,
          label: path
        };
      });
    }

    let nodes = _.map(links, (link, i) => {
      if (link.href) {
        return <li key={i}><Link to={link.href}>{link.label}</Link></li>;
      } else {
        return <li key={i}>{toTitleCase(link.label)}</li>;
      }
    });

    return (
      <ul className="breadcrumbs">
        <li><Link to={base}>{toTitleCase(type)}</Link></li>
        {nodes}
      </ul>
    );
  }
}

Breadcrumbs.propTypes = {
  splat: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired
};
