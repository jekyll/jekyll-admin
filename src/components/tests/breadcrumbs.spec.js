import React from 'react';
import { Link } from 'react-router';
import expect from 'expect';
import { mount } from 'enzyme';
import { capitalize } from '../../utils/helpers';
import moment from 'moment';
import Breadcrumbs from '../Breadcrumbs';

import { content } from './fixtures';

function setup(defaultContent = content) {
  let actions = {
    onChange: expect.createSpy()
  };
  const link = `/collections/${defaultContent.collection}`;
  const type = defaultContent.collection;

  let component = mount(
    <Breadcrumbs
      link={link}
      type={type}
      path={defaultContent.path}
      {...actions}/>
  );

  return {
    component: component,
    link: component.find(Link),
    li: component.find('li'),
    input: component.find('input'),
    actions
  };
}

describe('Components::Breadcrumbs', () => {
  it('should render correctly', () => {
    const { component, link, li, input } = setup();
    expect(link.length).toBe(1);
    expect(
      link.first().prop('children')
    ).toBe(capitalize(content.collection));
    expect(input.prop('defaultValue')).toBe(content.path);
  });
  it('should prepend date to input value/placeholder for new post', () => {
    const { input } = setup({ link:'test', collection:'posts', path:'' });
    const expectedValue = moment().format('YYYY-MM-DD') + '-your-title.md';
    expect(input.prop('placeholder')).toBe(expectedValue);
  });
  it('should call onChange', () => {
    const { input, actions } = setup();
    input.simulate('change');
    expect(actions.onChange).toHaveBeenCalled();
  });
});
