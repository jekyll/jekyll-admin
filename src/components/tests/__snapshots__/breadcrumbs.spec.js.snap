// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components::Breadcrumbs renders correctly 1`] = `
<ul
  className="breadcrumbs"
>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      Posts
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      1
    </a>
  </li>
</ul>
`;

exports[`Components::Breadcrumbs renders correctly 2`] = `
<ul
  className="breadcrumbs"
>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      Pages
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      1
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      2
    </a>
  </li>
</ul>
`;

exports[`Components::Breadcrumbs renders correctly 3`] = `
<ul
  className="breadcrumbs"
>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      Data Files
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      1
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      2
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      3
    </a>
  </li>
</ul>
`;

exports[`Components::Breadcrumbs renders correctly 4`] = `
<ul
  className="breadcrumbs"
>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      Drafts
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      1
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      2
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      3
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      4
    </a>
  </li>
</ul>
`;

exports[`Components::Breadcrumbs renders correctly 5`] = `
<ul
  className="breadcrumbs"
>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      Collections
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      1
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      2
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      3
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      4
    </a>
  </li>
  <li>
    <a
      onClick={[Function]}
      style={Object {}}
    >
      5
    </a>
  </li>
</ul>
`;
